### PR TITLE
ResourceNavigationLinks on subnets + Test-AzureRmPrivateIpAddress cmdlet 

### PR DESCRIPTION
**Pull request by AhmedSalama-MS**
## Comments

---

This checklist is used to make sure that common issues in a pull request are covered by the creator. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270%28v=vs.85%29.aspx).

Below in **Overall Changes**, check off the boxes that apply to your PR. For the categories that you did not check off, you can remove them from this body. Within each of the categories that you did select, make sure that you can check off **all** of the boxes. 

For information on cleaning up the commits in your pull request, click [here](../documentation/cleaning-up-commits.md).
## Overall Changes
- [x] [**MANDATORY** - General changes](#general)
- [x] [**MANDATORY** - Add/remove/edit test(s)](#tests)
- [x] [Add/remove/edit cmdlet(s)](#cmdlet-signature)
- [ ] [Add/remove/edit parameter(s)](#parameters)
- [ ] [Edit pipeline parameters](#parameters-and-the-pipeline)
### General
- [ ] Title of the PR is clear and informative
- [ ] There are a small number of commits that each have an informative message
- [ ] If it applies, references the bug/issue that the PR fixes
- [ ] All files have the Microsoft copyright header
- [ ] Cmdlets refer to management libraries through nuget references - no dlls are checked in
- [ ] The PR does not introduce breaking changes (unless a major version change occurs in the assembly and module)
### Tests
- [ ] PR includes test coverage for the included changes
- [ ] Tests must use xunit, and should either use Moq to mock management client calls, or use the scenario test framework
- [ ] PowerShell scripts used in tests must not use hard-coded values for location
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for existing resources
- [ ] Tests should not use App.config files for settings
- [ ] Tests should use the built-in PowerShell functions for generating random names when unique names are necessary - this will store names in the test recording
- [ ] Tests should use Start-Sleep to pause rather than Thread.Sleep
### Cmdlet Signature
- [ ] Cmdlet name uses an approved PowerShell verb - use the enums for `VerbsCommon`, `VerbsCommunication`, `VerbsLifecycle`, `VerbsOther` whenever possible
- [ ] Cmdlet noun name uses the AzureRm prefix for management cmdlets, and the Azure prefix for data plane cmdlets
- [ ] Cmdlet specifies the `OutputType` attribute if any output is produced; if the cmdlet produces no output, it should implement a `PassThrough` parameter
- [ ] If the cmdlet makes changes or has side effects, it should implement `ShouldProcess` and have `SupportShouldProcess = true` specified in the cmdlet attribute. See a discussion about correct `ShouldProcess` implementation [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change)
- [ ] Cmdlets should derive from AzureRmCmdlet for management cmdlets, and AzureDataCmdlet for data cmdlets
- [ ] If multiple parameter sets are implemented, the cmdlet should specify a `DefaultParameterSetName` in its cmdlet attribute
### Parameters
- [ ] Cmdlets should have no more than four positional parameters
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet
- [ ] Parameters should be explicitly marked as Mandatory or not, and should contain a HelpMessage
- [ ] No parameter is of type `object`.
  - Management cmdlets should have the following parameters and aliases:
    - ResourceGroupName with (optional) alias to `ResourceGroup` type string marked as [ValueFromPipelineByPropertyName]
    - Name with alias to `ResourceName` type string marked as [ValueFromPipelineByPropertyName]
    - Location (if appropriate) type string
    - Tag, type `HashTable`
### Parameters and the Pipeline
- [ ] Complex parameters should take values from the pipeline when possible, and certainly when they match the output type of another cmdlet
- [ ] Only one parameter should use ValueFromPipeline per parameter set; parameters from different parameter sets may have this attribute, but should not be convertible
- [ ] No parameter is of type `object`
- [ ] Each management cmdlet should have a parameter set that takes `ResourceGroupName` and `Name` from the pipeline by property value
- [ ] For a given resource type, it should be possible to pipe the output of `Get` and `New` cmdlets to the input of `Set`, `Update`, `Remove` and other action cmdlets for that resource

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@AhmedSalama-MS__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (ahsalam). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1 [hunk 1]
```diff
@@ -60,6 +60,15 @@ function Test-VirtualNetworkCRUD
         Assert-AreEqual $subnetName $list[0].Subnets[0].Name
         Assert-AreEqual "10.0.1.0/24" $list[0].Subnets[0].AddressPrefix
         Assert-AreEqual $expected.Etag $list[0].Etag
+
+        # Test virtual network private ip address - available
+        $testResponse1 = Get-AzureRmvirtualNetwork -Name $vnetName -ResourceGroupName $rgname | Test-AzureRmPrivateIpAddressAvailability -IpAddress "10.0.1.10"
+        Assert-AreEqual true $testResponse1.Available
+
+		# Test virtual network private ip address - not available
```
Comment (DeepakRajendranMsft) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1]: nit: tab on the comment

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1 [hunk 2]
```diff
@@ -265,4 +274,73 @@ function Test-VirtualNetworkPeeringCRUD
...
+		# Create redis-cache
+		$cacheCreated = New-AzureRmRedisCache -ResourceGroupName $rgname -Name $cacheName -Location $location -Size P1 -Sku Premium -SubnetId $subnet.Id
+
+		# In loop to check if cache exists
+		for ($i = 0; $i -le 60; $i++)
+		{
+		    Start-TestSleep 30000
+		    $cacheGet = Get-AzureRmRedisCache -ResourceGroupName $rgname -Name $cacheName
+		    if ([string]::Compare("succeeded", $cacheGet[0].ProvisioningState, $True) -eq 0)
+		    {
```
Comment (DeepakRajendranMsft) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1]: have the same indentation on all lines . select all -> edit -> untabify

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1 [hunk 3]
```diff
@@ -60,6 +60,15 @@ function Test-VirtualNetworkCRUD
...
         Assert-AreEqual $expected.Etag $list[0].Etag
+
+        # Test virtual network private ip address - available
+        $testResponse1 = Get-AzureRmvirtualNetwork -Name $vnetName -ResourceGroupName $rgname | Test-AzureRmPrivateIpAddressAvailability -IpAddress "10.0.1.10"
+        Assert-AreEqual true $testResponse1.Available
+
+		# Test virtual network private ip address - not available
+        $testResponse2 = Get-AzureRmvirtualNetwork -Name $vnetName -ResourceGroupName $rgname | Test-AzureRmPrivateIpAddressAvailability -IpAddress "10.0.1.3"
+        Assert-AreEqual false $testResponse2.Available
+		Assert-AreEqual 5 @($testResponse2.AvailableIpAddresses).Count
```
Comment (DeepakRajendranMsft) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1]: dont you want to add a test that checks ip address in use. You can use any of the nic test and add this in

File: src/ResourceManager/Network/Commands.Network/Models/PSPrivateIpAddressAvailabilityResponse.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Commands.Network.Models
+{
+    public class PSPrivateIpAddressAvailabilityResponse
+    {
+        public bool Available { get; set; }
+
+        public List<string> AvailableIpAddresses { get; set; }
```
Comment (DeepakRajendranMsft) [inline src/ResourceManager/Network/Commands.Network/Models/PSPrivateIpAddressAvailabilityResponse.cs]: you would need to add a serialization function like AvailableIpAddressesText. Check how the List<PSSubnet> is modelled in PSVirtualNetwork. 

Also rename to AvailableIPAddresses.

Also you would need to add this class to format.xml

File: src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
+﻿// ----------------------------------------------------------------------------------
```
Comment (DeepakRajendranMsft) [inline src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs]: rename to TestAzurePrivateIPAddressAvailabilityCmdlet

File: src/ResourceManager/Network/Commands.Network/Models/PSPrivateIpAddressAvailabilityResponse.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Commands.Network.Models
+{
+    public class PSPrivateIpAddressAvailabilityResponse
```
Comment (DeepakRajendranMsft) [inline src/ResourceManager/Network/Commands.Network/Models/PSPrivateIpAddressAvailabilityResponse.cs]: rename to PSPrivateIPAddressAvailability

File: src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        [Parameter(
+            Mandatory = true,
+            HelpMessage = "The Private Ip Address")]
+        [ValidateNotNullOrEmpty]
+        public string IpAddress { get; set; }
+
+        public override void Execute()
+        {
+            base.Execute();
+            var result = this.NetworkClient.NetworkManagementClient.VirtualNetworks.CheckIPAddressAvailability(this.VirtualNetwork.ResourceGroupName, this.VirtualNetwork.Name, this.IpAddress);
```
Comment (DeepakRajendranMsft) [inline src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs]: The output type here would be PrivateIPAddressAvailabilityResponse, no? (i.e) the object in SDK.

After the GET, you would have to covert to a PSPrivateIPAddressAvailabilityResponse (using the mapper class). See how other GET cmdlets work.

File: src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            Mandatory = true,
+            ValueFromPipeline = true,
+            HelpMessage = "The virtualNetwork")]
+        public PSVirtualNetwork VirtualNetwork { get; set; }
+
+        [Parameter(
+            Mandatory = true,
+            HelpMessage = "The Private Ip Address")]
+        [ValidateNotNullOrEmpty]
+        public string IpAddress { get; set; }
```
Comment (DeepakRajendranMsft) [inline src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs]: IPAddress

File: src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsDiagnostic.Test, "AzureRmPrivateIpAddressAvailability"), OutputType(typeof(PSPrivateIpAddressAvailabilityResponse))]
+    public class TestAzurePrivateIpAddressAvailabilityCmdlet : VirtualNetworkBaseCmdlet
+    {
+        [Parameter(
+            Mandatory = true,
+            ValueFromPipeline = true,
+            HelpMessage = "The virtualNetwork")]
+        public PSVirtualNetwork VirtualNetwork { get; set; }
```
Comment (DeepakRajendranMsft) [inline src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs]: You should have a parameter set to also accept the virtual network name and resource group name.

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1 (hunk 3 above)
Comment (DeepakRajendranMsft) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1]: Perhaps just remove this change and create a new test with a nic

File: src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj [hunk 10]
```diff
@@ -63,8 +63,16 @@
     <Reference Include="Microsoft.Azure.Management.Authorization">
       <HintPath>..\..\..\packages\Microsoft.Azure.Management.Authorization.2.0.0\lib\net40\Microsoft.Azure.Management.Authorization.dll</HintPath>
     </Reference>
-    <Reference Include="Microsoft.Azure.Management.Network, Version=6.0.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.6.0.3-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.Network">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.6.1.0-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
```
Comment (DeepakRajendranMsft) [inline src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj]: use: 6.1.1 a new version has been added

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1 (hunk 2 above)
Comment (AhmedSalama-MS) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1]: Fixed.

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1 (hunk 1 above)
Comment (AhmedSalama-MS) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1]: Fixed

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1 (hunk 3 above)
Comment (AhmedSalama-MS) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/VirtualNetworkTests.ps1]: Shouldn't make much difference to check in use nic or reserved ip address. Will leave this as is.

File: src/ResourceManager/Network/Commands.Network/Models/PSPrivateIpAddressAvailabilityResponse.cs (hunk 6 above)
Comment (AhmedSalama-MS) [inline src/ResourceManager/Network/Commands.Network/Models/PSPrivateIpAddressAvailabilityResponse.cs]: Fixed

File: src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs (hunk 7 above)
Comment (AhmedSalama-MS) [inline src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs]: Fixed

File: src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs (hunk 5 above)
Comment (AhmedSalama-MS) [inline src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs]: Fixed

File: src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs (hunk 8 above)
Comment (AhmedSalama-MS) [inline src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs]: Fixed

File: src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs (hunk 9 above)
Comment (AhmedSalama-MS) [inline src/ResourceManager/Network/Commands.Network/VirtualNetwork/TestAzurePrivateIpAddressAvailabilityCmdlet.cs]: Fixed

**Comment by cormacpayne** [comment]
@AhmedSalama-MS Hey Ahmed, are these changes trying to make it into the September release? If so, please change the base branch to release-2.1.0

**Comment by AhmedSalama-MS** [comment]
Leave a comment

**Pull request closed without merging**